query_optimizer/rules/UnnestSubqueries.cpp

```diff
@@ -0,0 +1,720 @@
+        if (new_operand != operand && !has_change) {
+          has_change = true;
+        }
+        DCHECK(new_operand != nullptr);
+        new_operands.emplace_back(std::static_pointer_cast<const E::Predicate>(new_operand));
+      }
+
+      if (has_change) {
+        return E::LogicalOr::Create(new_operands);
+      }
+      return node;
+    }
+    default:
+      std::vector<E::ExpressionPtr> new_children;
+      bool has_change = false;
+      for (const E::ExpressionPtr &child : node->children()) {
+        const E::ExpressionPtr new_child =
+            applyInternal(allow_exists_or_in, child);
+        if (new_child != child && !has_change) {
+          has_change = true;
+        }
+        DCHECK(child != nullptr);
+        new_children.emplace_back(new_child);
+      }
+
+      if (has_change) {
+        return node->copyWithNewChildren(new_children);
+      }
+      return node;
+  }
+}
+
+void UnnestSubqueriesForExpession::transformExists(
+    const E::Exists *exists_predicate) {
+  std::vector<E::AttributeReferencePtr> probe_join_attributes;
+  std::vector<E::AttributeReferencePtr> build_join_attributes;
+  std::vector<E::PredicatePtr> non_hash_join_predicates;
+  UnnestSubqueriesForNonRootLogical unnest_logical_rule(false,  // scalar_query
+                                                        visible_attributes_from_outer_query_,
+                                                        context_,
+                                                        uncorrelated_subqueries_,
+                                                        &probe_join_attributes,
+                                                        &build_join_attributes,
+                                                        &non_hash_join_predicates);
+  DCHECK_EQ(probe_join_attributes.size(), build_join_attributes.size());
+  const L::LogicalPtr new_subquery =
+      unnest_logical_rule.apply(exists_predicate->exists_subquery()->subquery());
+
+  if (probe_join_attributes.empty()) {
+    if (!non_hash_join_predicates.empty()) {
+      THROW_SQL_ERROR() << "Correlated queries must have an equality join predicate with the outer query";
+    }
+    THROW_SQL_ERROR() << "EXISTS subquery cannot be un-correlated with the outer query";
+  }
+
+  correlated_query_info_vec_->emplace_back(CorrelatedQueryInfo::kSemiJoin,
+                                           new_subquery,
+                                           std::move(probe_join_attributes),
+                                           std::move(build_join_attributes),
+                                           std::move(non_hash_join_predicates));
+}
+
+E::ExpressionPtr DeOuterAttributeReference::applyToNode(const E::ExpressionPtr &input) {
+  E::AttributeReferencePtr attr;
+  if (E::SomeAttributeReference::MatchesWithConditionalCast(input, &attr)
+      && attr->scope() == E::AttributeReferenceScope::kOuter
+      && uncorrelated_subqueries_.find(attr->id()) == uncorrelated_subqueries_.end()) {
```

Comment:
As a convention in our codebase, we put `&&` or `||` at the end of line.

So it looks like:

```
  if (E::SomeAttributeReference::MatchesWithConditionalCast(input, &attr) &&
      attr->scope() == E::AttributeReferenceScope::kOuter &&
      uncorrelated_subqueries_.find(attr->id()) == uncorrelated_subqueries_.end()) {
```